apps/settings/js/modules/feature_model.js

```diff
@@ -0,0 +1,108 @@
+/**
+ * A model needs to be created with an options object with the following
+ * three functions:
+ *
+ * onInit is called once with a function as the first argument when the
+ * model is initializing. The function is used for reporting the state
+ * changes. The function should return a initial state immediately.
+ *
+ * onSetState is called with a desired state. Do the operation required for
+ * switching to the desired state and return a promise wrapping the state
+ * after the operation no matter it resolves or rejects.
+ *
+ * onGetState is called for requesting the current state.
+ */
+define(function(require) {
+  'use strict';
+
+  var Module = require('modules/base/module');
+  var Observable = require('modules/mvvm/observable');
+
+  /**
+   * @class FeatureModel
+   * @param {Function} onInit
+   * @param {Function} onSetState
+   * @param {Function} onGetState
+   * @returns {FeatureModel}
+   */
+  var FeatureModel = Module.create(
+    function FeatureModel({onInit, onGetState, onSetState}) {
+      if (typeof onInit !== 'function' ||
+          typeof onGetState !== 'function' ||
+          typeof onSetState !== 'function') {
+        this.throw('invalid arguments');
+      }
+      this.super(Observable).call(this);
+
+      var argument = arguments[0];
+      this._init(onInit.bind(argument), onGetState.bind(argument),
+        onSetState.bind(argument));
+  }).extend(Observable);
+
+  Observable.defineObservableProperty(FeatureModel.prototype, 'targetState', {
+    value: null
+  });
+
+  Observable.defineObservableProperty(FeatureModel.prototype, 'currentState', {
+    readonly: true,
+    value: null
+  });
+
+  Observable.defineObservableProperty(FeatureModel.prototype, 'transitioning', {
+    readonly: true,
+    value: false
+  });
+
+  FeatureModel.prototype._init = function(onInit, onGetState, onSetState) {
+    this._onInit = onInit;
+    this._onGetState = onGetState;
+    this._onSetState = onSetState;
+
+    this.targetState = this._onInit((realState) => {
```

Comment:
You can consider this is the defined interface between FeatureModel and its option. The option provides an interface comprising of `onInit`, `onGetState`, `onSetState` to FeatureModel. The Feature provides necessary information to the option.
An analogy would be a button passing an event object to its handler.